CASSANALYTICS-49: Support UDTs inside collections

skoppu22 opened the pull request:
Currently Cassandra bulk writer not properly supporting UDTs inside collections. Also complex structures like collections inside UDTs are not supported properly. Adding support for these cases.

Note: Tuples with UDTs are not yet supported properly, it will be taken care in another Jira/PR

Circle CI passed: https://app.circleci.com/pipelines/github/skoppu22/cassandra-analytics/19/workflows/187a5c65-b03f-495e-b29b-914dd0483642

skoppu22 committed:
Support UDTs inside collections

test build

checkstyle

yifan-c commented on cassandra-analytics-integration-framework/src/main/java/org/apache/cassandra/testing/TestUtils.java:
```diff
@@ -36,7 +36,7 @@ public final class TestUtils
     private static final AtomicInteger TEST_TABLE_ID = new AtomicInteger(0);
 
     //    public static final int ROW_COUNT = 10_000;
-    public static final int ROW_COUNT = 1_000;
+    public static final int ROW_COUNT = 100;
```
can you revert the change? I assume you lower the row count for local testing.

skoppu22 replied:
Some tests were timing out when ran on local, especially when ran from Intellij, so had to reduce it. Can we just continue with lesser number of rows so we don't need to change this on local everytime we run?

yifan-c replied:
I ran the tests after reverting `ROW_COUNT`, cannot reproduce timing out. Maybe it was related to the deep copy for all columns (which is now resolved). Can you revert the change?

frankgh replied:
this change should be reverted. Are you seeing issues after the new change?

skoppu22 replied:
I am seeing quite slower runs, sometimes timeouts only when running under Intellij. Reverted this change.

yifan-c commented on cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RecordWriter.java:
```diff
@@ -389,19 +391,52 @@ private Map<String, Object> getBindValuesForColumns(Map<String, Object> map, Str
 
     private Object maybeConvertUdt(Object value)
     {
+        if (value instanceof List)
+        {
+            List<Object> resultList = new ArrayList<>();
+            for (Object entry : (List<?>) value)
+            {
+                resultList.add(maybeConvertUdt(entry));
+            }
+
+            return resultList;
+        }
+
+        if (value instanceof Set)
+        {
+            Set<Object> resultList = new HashSet<>();
+            for (Object entry : (Set<?>) value)
+            {
+                resultList.add(maybeConvertUdt(entry));
+            }
+
+            return resultList;
+        }
+
+        if (value instanceof Map)
+        {
+            Map<Object, Object> resultMap = new HashMap<>();
+            for (Map.Entry<?, ?> entry : ((Map<?, ?>) value).entrySet())
+            {
+                resultMap.put(maybeConvertUdt(entry.getKey()), maybeConvertUdt(entry.getValue()));
+            }
+
+            return resultMap;
+        }
```
those 3 branches will copy the list unnecessarily if they do _not_ contain UDT in the table schema. It would produce a lot of short-lived objects. Note that there are millions or more rows to write. 

One idea to skip the unnecessary conversion is to check the table schema. 
1. `RecordWriter` has access to the table schema, e.g. `CqlTable`. 
2. `CqlTable` can mark the columns where contain UDT. 
3. Instead of calling `maybeConvertUdt` on every column in each row, first check whether `maybeConvertUdt` should be invoked for this column. Now, since you know which columns contain UDT, you can even drop "maybe" from the method name. It would clarify the code too.

skoppu22 replied:
- Updated CqlTable to maintain columns with UDTs and check it before calling maybeConvertUdt
- There’s still a chance of creating unnecessary lists, sets, or maps, like udt<list<udt<set<map>>>>. We create new sets and maps without knowing if there are more UDTs inside. I tried counting UDTs in a column (inside determineColumnsWithUdts) to avoid this, but it may not work for all cases. It doesn’t differentiate when the same UDT type is used multiple times, like in map<udtType1, udtType1>, where the UDT count is 1 instead of 2. This is fine for Maps, but it could break things in non-Map cases.

yifan-c replied:
having columns referrencing UDTs flagged is good enough. Thank you for adding that!

yifan-c commented on cassandra-analytics-integration-tests/src/test/java/org/apache/cassandra/analytics/BulkWriteUdtTest.java:
```diff
@@ -88,17 +141,298 @@ void testWriteWithNestedUdt()
+        cluster.schemaChangeIgnoringStoppedInstances(String.format("insert into %s (id, udtset) values (%d, " +
+                        "{{f1:null, f2:null, f3:null}})",
+                SET_OF_UDT_SOURCE_TABLE, i++));
+
```
`schemaChangeIgnoringStoppedInstances` is to change schema. It is not a proper use of the api to insert data. 

I think you want to use `cluster.coordinator(1).executeWithResult(QUERY)` instead.

skoppu22 replied:
- schemaChangeIgnoringStoppedInstances also uses executeWithResult internally, but it uses Monitor additionally. 
- Changed to use execute for insert and select queries

yifan-c commented on cassandra-analytics-integration-tests/src/test/java/org/apache/cassandra/analytics/BulkWriteUdtTest.java:
```diff
@@ -32,7 +32,7 @@
 import org.jetbrains.annotations.NotNull;
 
-import static org.apache.cassandra.testing.TestUtils.DC1_RF3;
+import static org.apache.cassandra.testing.TestUtils.DC1_RF1;
```
Curious about the reason of the change? If unnecessary, please revert.

skoppu22 replied:
This should be reverted, was needed only for local testing

skoppu22 replied:
Done

yifan-c commented on cassandra-analytics-integration-tests/src/test/java/org/apache/cassandra/analytics/BulkWriteUdtTest.java:
```diff
@@ -59,6 +59,59 @@ class BulkWriteUdtTest extends SharedClusterSparkIntegrationTestBase
+    public static final String UDT_WITH_COLLECTION_OF_UDT_TABLE_CREATE = "CREATE TABLE %s.%s (\n"
+            + "            id BIGINT PRIMARY KEY,\n"
+            + "            outerudt frozen<%s>)";
+
+
```
There is no test covering the scenario of UDTs in tuple. Are you going to update this patch or address in a follow-up? (I am OK with either of them)

skoppu22 replied:
As I mentioned in the PR description, there is lot more to be done to handle Tuples, support for Tuples will come in another PR/task

yifan-c reviewed (commented):
Sorry for the delay. I thought I have reviewed, but apparently not :p

skoppu22 committed:
optimize collections creation

yifan-c committed:
ninja; fix typo in LICENSE.txt

bbotella committed:
CASSANALYTICS-2: Updates to latest gradle version (#96)

Patch by Bernardo Botella; Reviewed by Yifan Cai for CASSANALYTICS-2

anderoo committed:
CASSANALYTICS-50: Add support for vnodes (#93)

Patch by Andrew Johnson; reviewed by James Berragan, Yifan Cai, Francisco Guerrero for CASSANALYTICS-50

skoppu22 committed:
update udt checks

skoppu22 committed:
Merge branch 'apache:trunk' into udtcollections

skoppu22 committed:
use execute

yifan-c commented on cassandra-analytics-common/src/main/java/org/apache/cassandra/spark/data/CqlTable.java:
```diff
@@ -50,6 +51,7 @@ public class CqlTable implements Serializable
     private final Set<CqlField.CqlUdt> udts;
 
     private final Map<String, CqlField> fieldsMap;
+    private final Map<String, Boolean> columnsWithUdts;
```
👍 
An minor improvement is to a `Set` instead. Only the columns referencing UDT are included in the set.

yifan-c commented on cassandra-analytics-common/src/main/java/org/apache/cassandra/spark/data/CqlTable.java:
```diff
@@ -242,6 +246,31 @@ public int indexCount()
+    /**
+     * Check each column of the table for UDT type somewhere nested inside it and maintain map of true/false accordingly
+     * @return map of column name to true/false
+     */
```
👍 thanks for adding javadocs

yifan-c commented on cassandra-analytics-common/src/main/java/org/apache/cassandra/spark/data/CqlTable.java:
```diff
@@ -242,6 +246,31 @@ public int indexCount()
+     * @param fieldName name of the column
+     * @return true if the column has UDT type , false otherwise
+     */
+    public boolean containsUdt(String fieldName)
```
nit: to be consistent with Cassandra and to better describe the column, how about naming it as `referencingUserType`?

skoppu22 replied:
We use the word UDT extensively all over the repo. Better to keep the name containsUdt to be inline with remaining code.

yifan-c replied:
Sounds good. Given that you added the detailed javadoc (awesome!), I have no concern.

yifan-c commented on cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RecordWriter.java:
```diff
@@ -380,28 +382,69 @@ private Map<String, Object> getBindValuesForColumns(Map<String, Object> map, Str
         return map;
     }
 
     private Object maybeConvertUdt(Object value)
```
It would be nice to add java doc to describe the method, e.g. why it checks for collections in the implementation.

yifan-c commented on cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RecordWriter.java:
```diff
@@ -380,28 +382,69 @@ private Map<String, Object> getBindValuesForColumns(Map<String, Object> map, Str
+
+        if (value instanceof Set)
+        {
+            Set<Object> resultList = new HashSet<>();
```
nit: `resultSet`?

yifan-c commented on cassandra-analytics-integration-tests/src/test/java/org/apache/cassandra/analytics/BulkWriteUdtTest.java:
```diff
@@ -59,6 +60,61 @@ class BulkWriteUdtTest extends SharedClusterSparkIntegrationTestBase
                                                      + "           nested " + NESTED_FIELD_UDT_NAME + ");";
 
+    // UDT with list, set and map in it
+    public static final String UDT_WITH_COLLECTIONS_TYPE_NAME = "udt_with_collections";
+    public static final String UDT_WITH_COLLECTIONS_TYPE_CREATE = "CREATE TYPE " + TEST_KEYSPACE + "." + UDT_WITH_COLLECTIONS_TYPE_NAME +
+            " (f1 list<text>, f2 set<int>, f3 map<int, text>);";
+
```
It would be nice to use QT to fuzz test the different combinations of types. Just say it out load, no need to address.

frankgh commented on cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RecordWriter.java:
```diff
@@ -380,28 +382,69 @@ private Map<String, Object> getBindValuesForColumns(Map<String, Object> map, Str
         for (int i = 0; i < columnNames.length; i++)
         {
-            map.put(columnNames[i], maybeConvertUdt(values[i]));
+            if (cqlTable().containsUdt(columnNames[i]))
```
can we extract the method invocation for `cqlTable` and assign it to a variable outside of the for loop?

skoppu22 replied:
cqlTable() already uses a class local variable `cqlTable`, populates it only once if it is not set and references the same in subsequent invocations. Using a local variable here in this caller only avoids the function call, but compiler must be smart enough to avoid subsequent calls I believe. Also `cqlTable` is volatile, there must be a reason for making it volatile, so better we reference the same. Infact we can reference that class local variable `cqlTable` directly here by calling cqlTable() only once in the class init, but I wouldn't do that without understanding why it was implemented this way previously.

skoppu22 replied:
I see @yifan-c added cqlTable. @yifan-c Can we call cqlTable() only once in the class init and reference the class local variable in rest of the class? I see `cqlTable` is volatile, but it is assigned only once when it is null. Does it need to be really volatile?

frankgh replied:
I think we want to avoid having a call to a method in a tight loop. So it's better to get a reference of this variable before the loop

yifan-c replied:
> Infact we can reference that class local variable cqlTable directly here by calling cqlTable() only once in the class init
This sounds good. Given that the writer requires cqlTable to be initialized, the initialization can be done eagerly.

frankgh commented on cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RecordWriter.java:
```diff
@@ -380,28 +382,69 @@ private Map<String, Object> getBindValuesForColumns(Map<String, Object> map, Str
     {
+        if (value instanceof List)
+        {
+            List<Object> resultList = new ArrayList<>();
```
minor, avoid creating a list when the source is empty. ie. if !value.isEmtpy() -> process

frankgh commented on cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RecordWriter.java:
```diff
@@ -380,28 +382,69 @@ private Map<String, Object> getBindValuesForColumns(Map<String, Object> map, Str
+            return resultList;
+        }
+
+        if (value instanceof Set)
```
minor , avoid creating a set when source set is empty

frankgh commented on cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RecordWriter.java:
```diff
@@ -380,28 +382,69 @@ private Map<String, Object> getBindValuesForColumns(Map<String, Object> map, Str
+            return resultList;
+        }
+
+        if (value instanceof Map)
```
minor avoid creating a map when the source map is empty

frankgh commented on cassandra-analytics-integration-tests/src/test/java/org/apache/cassandra/analytics/SharedClusterSparkIntegrationTestBase.java:
```diff
@@ -171,28 +173,35 @@ public void checkSmallDataFrameEquality(Dataset<Row> expected, Dataset<Row> actu
+        Set<String> driverEntries = new HashSet<>();
+        driverData.forEach(row -> driverEntries.add(driverRowFormatter
+                .apply(row)
+                // Driver Codec writes "NULL" for null value. Spark DF writes "null".
+                .replace("NULL", "null")
+                // driver writes lists as [] and sets as {},
+                // whereas spark entries have the same type WrappedArray for both lists and sets
+                .replace('[', '{')
+                .replace(']', '}')));
```
I think this logic should be encapsulated in the function itself, and not here.

skoppu22 replied:
Which function do you mean? We repeat the same logic for all driver entries formatters for UDTs (and also Tuples which is coming up today). So keeping it in one place instead of repeating the same in each formatter.

skoppu22 replied:
Moved out collections specific formatting to corresponding formatter, to avoid other tests getting impacted by chance

skoppu22 committed:
minor comments

yifan-c commented:
Please update `CHANGES.txt`

skoppu22 committed:
add change record

frankgh commented on CHANGES.txt:
```diff
@@ -1,4 +1,5 @@
 1.0.0
+ * Bulk writer needs to support collections with UDTs
```
can we add the ticket here? 
```suggestion
 * Bulk writer needs to support collections with UDTs (CASSANALYTICS-49)
```

frankgh reviewed (approved):
+1 LGTM

yifan-c committed:
Merge branch 'trunk' into udtcollections
